[PVT#180979763] Add placement constraint for current capacity provider AMI

andyzito opened the pull request:
[Pivotal story](https://www.pivotaltracker.com/story/show/180979763)

Next steps:

1. Deploy [Ansible change](https://github.com/greenriver/ansible-fresh/pull/66)m everywhere (since it doesn't break anything to have it in places it isn't needed yet).
   - [x] staging
   - [x] production
3. Apply [terraform change](https://github.com/greenriver/terraform/pull/80). This shouldn't affect either Warehouse or CAS until the application code which reads from the Parameter Store is deployed.
   - [x] Confirm that we have three separate capacity providers
   - [x] Confirm different AMI's for each one (use placeholder AMIs)
   - [x] Confirm that parameter store has the right name (blue)
   - [x] Replace placeholder AMIs with the real AMI (same everywhere for now) and reapply
   - [x] Merge PR
4. Deploy application code to QA. Confirm that all services/tasks are using the `blue` capacity provider.
   - [x] done?
6. Deploy the application code elsewhere. This should be safe to do over time and shouldn't require direct coordination (Elliot should be able to deploy on his own at whatever pace he prefers)
   - [x] done?
8. Repeat for CAS.
9. Clean up Terraform code (remove v2 providers) and apply.

Note: Will this mean three different fully sized autoscaling group? Or will the ones not being used scale down to nothing? Should confirm that before apply.

OLD:

This adds logic to the deployment scripts, cron installers, workoff arbiter, etc. to query the capacity provider for the current AMI ID and add that as a constraint to the deployed service or task. This is currently deployed to QA and you can see the following placement constraint shows up on the various services:
> **Constraint**   memberOf(attribute:ecs.ami-id == ami-0571c653798220ccd)

@schwartzkylej Might be worth our time to make a no-op change to the AMI so we can test this. I'm not 100% sure that the querying I wrote will reliably get the desired ID when we're in the middle of the AMI upgrade process.

@eanders I can see the constraints on the dj services, so we're good there. But we should also make double sure the cron tasks and workoff jobs are getting the constraint -- and it looks like both of those are spinning up tasks directly (rather than within a service). Is there a way for you to manually trigger a cron task and a workoff task? If there is, let's coordinate and I'll check the running task for the constraint.

Relies on a change to the task policy which is deployed to QA, but will need to be deployed to all environments prior to deploying this code: https://github.com/greenriver/ansible-fresh/pull/66

Commit:
Add AMI placement constraint to all services

Commit:
Shared AwsSdkMethods module

Commit:
Woops, singleton_methods don't work that way

Commit:
More streamlining

Commit:
Gem elasticloadbalancingv2

Commit:
Fix cluster_name env var chain

Commit:
Final value in the ENV.fetch chain

Commit:
.present? is Railsy and not available in some places

Commit:
....oh

Commit:
Fixing hash .find syntax

Commit:
Still fixing find syntax...

Commit:
Fix getting first item in hash

Commit:
[screaming] I forgot to remove byebug

schwartzkylej commented:
Added @yakloinsteak to the review. Happy to approve once he has eyes on it.

eanders commented on app/jobs/workoff_arbiter.rb:
```diff
@@ -1,3 +1,5 @@
+require_relative '../../config/deploy/docker/lib/aws_sdk_helpers'
```
This needs to move below the copyright.  We have scripts for updating it, but only if it starts at the top of the file and is followed by an empty line.

eanders reviewed (commented):
@andyzito , can you make sure you run `overcommit --install` in the warehouse directory?  We have enforced styles, but it looks like you may not have the rubocop linting running.

eanders commented on config/deploy/docker/lib/aws_sdk_helpers.rb:
```diff
@@ -0,0 +1,105 @@
+
+
+    def _spot_capacity_provider_name
+      _capacity_providers.find { |k,v| v[:name].match(/spt-v2/) }[1][:name]
+    end
+
+    def _on_demand_capacity_provider_name
+      _capacity_providers.find { |k,v| v[:name].match(/ondemand-v2/) }[1][:name]
```
Should these pull the `v2` out of an environment variable?  Seems like one more place we'll need to keep track of a version.  I'd rather see it in the environment, or at least in a constant.

yakloinsteak replied:
I like that @eanders. Let's not block this PR, but I made https://www.pivotaltracker.com/story/show/181282309.

eanders commented:
@andyzito , the cron containers are kicked off via ECS scheduled tasks, there should be one every 2 minutes to check to see if we need to spin up workoff containers.  It may be worth adding a new version of `app/jobs/test_job.rb` that is queued as `queue_as ENV.fetch('DJ_LONG_QUEUE_NAME', :long_running)` which will mean it gets picked up by a work-off worker.  It could be really simple, just a sleep(500) or something so that it takes long enough to run for you to see it in the ECS console or API.

yakloinsteak commented on config/deploy/docker/lib/aws_sdk_helpers.rb:
```diff
@@ -0,0 +1,105 @@
+require 'aws-sdk-cloudwatch'
+require 'aws-sdk-cloudwatchlogs'
+
+module AwsSdkHelpers
```
Can we use ActiveSupport::Concern for this module?

yakloinsteak commented on config/deploy/docker/lib/aws_sdk_helpers.rb:
```diff
@@ -0,0 +1,105 @@
+    define_method(:cloudwatchevents) { AwsSdkHelpers::ClientMethods.cloudwatchevents }
+    define_method(:cw)               { AwsSdkHelpers::ClientMethods.cw }
+    define_method(:cwl)              { AwsSdkHelpers::ClientMethods.cwl(
+      profile: respond_to?(:_cluster_name) ? _cluster_name : AwsSdkHelpers::Helpers.cluster_name,
```
What's the profile parameter for? There might have been some old code using profiles when instantiating clients, but I don't think we want them anymore. I'd remove it wherever you see it unless I'm missing something.

yakloinsteak commented on config/deploy/docker/lib/aws_sdk_helpers.rb:
```diff
@@ -0,0 +1,105 @@
+    include AwsSdkHelpers::ClientMethods
+
+    def self.cluster_name
+      ENV.fetch('CLUSTER_NAME', ENV.fetch('AWS_CLUSTER', ENV.fetch('AWS_PROFILE', ENV.fetch('AWS_VAULT', ''))))
```
We shouldn't propagate my bad defaulting logic if possible. We should just settle on one of these (I like the first).

yakloinsteak commented on config/deploy/docker/lib/aws_sdk_helpers.rb:
```diff
@@ -0,0 +1,105 @@
+
+    def self.capacity_providers(cluster)
+      r = {}
+      capacity_provider_names = AwsSdkHelpers::ClientMethods.ecs.describe_clusters(clusters: [cluster]).clusters.first.capacity_providers
```
I think using a concern like I mentioned above will let us not need the module namespacing here.

andyzito replied:
It didn't seem happy without the namespacing. I don't mind the explicit class reference myself.

yakloinsteak commented on config/deploy/docker/lib/aws_sdk_helpers.rb:
```diff
@@ -0,0 +1,105 @@
+    def self.capacity_providers(cluster)
+      r = {}
+      capacity_provider_names = AwsSdkHelpers::ClientMethods.ecs.describe_clusters(clusters: [cluster]).clusters.first.capacity_providers
+      AwsSdkHelpers::ClientMethods.ecs.describe_capacity_providers(capacity_providers: capacity_provider_names).capacity_providers.map do |capacity_provider|
```
I'd say to use `each` if we're not saving the result.

yakloinsteak commented on config/deploy/docker/lib/aws_sdk_helpers.rb:
```diff
@@ -0,0 +1,105 @@
+    def self.default_placement_constraints(ami_id:)
+      [
+        {
+          expression: "attribute:ecs.ami-id == #{ami_id}",
```
What's the plan for handling the situation right after we publish a new launch template? I think we're fine for everything already running when that moment happens, but I think scheduled tasks, new service deployments, and the occasional container crash will find the new containers have no place to go.

I think this constraint needs to be more complex (or the code above) so that we only constrain if there's actually ec2 instances running that could fulfill the request and that have capacity.

andyzito replied:
Added logic to query ec2.describe_instances for capacity provider instances w/matching ami.

andyzito replied:
@yakloinsteak @eanders We need to keep in mind here that, unless I'm misinterpreting the code, we're looking at these constraints only being guaranteed to be re-evaluated during deployment. I think that's fine, and the process would look something like:

- update ami
- update capacity_provider config to use new ami_id
- application now has a constraint on the old ami_id, and since those instances stick around while they're being used, we're in a safe place here
- deploy application, which will update the constraints to use the new ami_id (unless no instances w/that ami are available yet)
- now the application is only able to spin up new things on the new ami, and things start moving over

This isn't like, a problem, I just wanted to call out that the constraints aren't necessarily going to be re-evaluated every time, say, a new cron task is spun up.

yakloinsteak commented on config/deploy/docker/lib/deployer.rb:
```diff
@@ -51,7 +50,10 @@ class Deployer
+  attr_accessor :cluster
+
   def initialize(target_group_name:, assume_ci_build: true, secrets_arn:, execution_role:, task_role:, dj_options: nil, web_options:, registry_id:, repo_name:, fqdn:)
+    self.cluster           = ENV.fetch('AWS_CLUSTER') { ENV.fetch('AWS_PROFILE') { ENV.fetch('AWS_VAULT') } }
```
Yeah, let's simplify this if possible. Definitely remove the profile one.

yakloinsteak reviewed (changes requested):
Looking good so far. Thanks, @andyzito

Commit:
Simplify _cluster_name and other small fixes

Commit:
Require below copyright

Commit:
Syntax error

andyzito commented:
@eanders What do I do after the `overcommit --install` to get it to do the linting?

I will look at testing the cron/workoff jobs as you suggested later today.

Commit:
Add logic to ensure that there's at least one capacity provider instance for AMI before applying constraint

Commit:
Hmm, constraints aren't showing up on cron tasks...

Commit:
Still trying to get constraints to show up on cron tasks

Commit:
Add workoff test job

yakloinsteak commented on config/deploy/docker/lib/aws_sdk_helpers.rb:
```diff
@@ -0,0 +1,128 @@
+      }
+      matching_instances = ClientMethods.ec2.describe_instances(payload)
+
+      if matching_instances.reservations.count > 0 && !ami_id.empty?
```
This is better, but let's talk it through before we merge this. I still have this feeling we can end up with un-schedulable containers.

If you start deploying after the AMI change, the capacity provider might have enough capacity and not spin up a new instance, and if you start using instance refresh in the ASG, you're draining all the services, getting one new-ami instance at a time, but the services still want the old AMI.

Could we update the constraints as part of an instance refresh (little by little)?

My current thinking is that 1 new ami instance might not be enough and that we'll need a process to artificially expand the cluster temporarily (like with a bunch of nginx containers with medium sized RAM quotas). Each deployment after an update would retire some of the nginx containers.

I don't think this is the best idea, so let's talk to the group. Maybe I'm overthinking this.

Might have to take a step back and use a new capacity provider whenever we have a new AMI. That might simplify all of this. I think we might be finally seeing why they don't recommend constraints with capacity providers.

andyzito replied:
Sounds like we should definitely talk through this. One note is that I think if you start deploying after the AMI change and there are no new instances being spun up, then there will be no constraints applied to everything created during the deployment. That is, it won't keep the constraint w/the old AMI if the capacity provider is set for a new AMI. And it won't constrain based on that new AMI if there are no available instances.

yakloinsteak replied:
That sounds right.

andyzito commented:
Note: I haven't taken the time to figure out overcommit yet, but I ran the rubocop linting manually and there were no errors relating to my changes.

eanders commented on config/deploy/docker/lib/aws_sdk_helpers.rb:
```diff
@@ -0,0 +1,128 @@
+          },
+          {
+            name: "image-id",
+            values: [ami_id]
```
@andyzito , if rubocop is setup correctly (or overcommit) it would require things like final commas in hashes.  I'm happy to jump on a quick call to help get it setup.

Commit:
Fix rubocop

Commit:
Fix waaaaay more rubocop stuff

Commit:
So. Much. Rubocop.

Commit:
Get capacity provider names from parameter store

Commit:
.positive -> .positive?

Commit:
Output warning instead of dropping to byebug

andyzito commented:
For @eanders , we discussed this in the DevOps meeting and we'd like to alter our approach to use multiple capacity providers rather than mess around with ami constraints. The idea is to have two capacity provider sets (I chose red/green as the identifiers but we can change that to literally anything) for four total capacity providers (spot and on-demand * red and green). Then we store the current capacity provider names in the Parameter Store, and the ecs deploy code (and the in-application task launcher code like the workoff arbiter) will read the names from that parameter. When we update the AMI, we will do so in the not-in-use capacity provider set, and update the parameters to tell OP to switch over.

I've made some prospective terraform updates: https://github.com/greenriver/terraform/pull/80

And I've updated the code here to read the capacity provider names from the parameter store. You can see example parameters in the SSM console right now. I added the capability to override the `/Default/` parameters for any specific project by using the target group as the namespace (e.g. `/qa-warehouse-staging-ecs/CapacityProviders/Spot` instead of `/Default/CapacityProviders/Spot`). The code will automatically read from the one matching the project if it exists. Maybe that won't end up being needed but it felt good to have the option, especially as we experiment with this.

Commit:
Fix target_group_name determination when getting capacity provider

andyzito commented:
[moved to description]

yakloinsteak commented:
I believe that as the capacity moves to the new capacity provider, the "old" ASG will scale in to nothing. Keep in mind that when a new capacity provider/asg scales out, it will not be balanced correctly (e.g. all the initial containers will be on one host, not distributed across AZs).

Commit:
Change root namespace of paramter store paths

Commit:
Forgot the decryption...

Commit:
Remove placement_constraints

andyzito commented:
Status update: I've deployed this to QA and things look stable and good. I believe that this code is ready to merge and deploy elsewhere.

yakloinsteak commented on config/deploy/docker/lib/cron_installer.rb:
```diff
@@ -157,25 +145,22 @@ def get_cron_expression(line)
 
   def get_command(line)
-    reg_match = line.match(%r{/bin/bash -l -c '(.+)'$})
+    reg_match = line.match(/\/bin\/bash -l -c '(.+)'$/)
```
anti-lint. Love it.

yakloinsteak commented on config/deploy/docker/lib/aws_sdk_helpers.rb:
```diff
@@ -0,0 +1,112 @@
+        {
+          names: [
+            default_path,
+            namespaced_path,
```
Could you comment this method with how this defaulting logic works? Is this related to the transition and will become defunct afterwards? If that's the case, we might want to make a PT to clean this up afterwards.

andyzito replied:
For some added context:
- I'm taking this out because we don't actually need it right now.
- a more flexible, robust, and well-commented version of this functionality will be needed in ecs-deploy-public v2 (aka ecs-deploy-private), but we'll get to that later.

yakloinsteak reviewed (approved):
Looks good.

eanders committed:
Merge branch 'release-28' into stories/180979763-ami-placement-constraint

eanders committed:
Wait longer during deploys

Commit:
Merge branch 'release-28' into stories/180979763-ami-placement-constraint

Commit:
Remove namespace/fallback logic for parameters (we don't actually need it right now)
